lib/grid_search.py

```diff
@@ -42,18 +42,46 @@ def train(self):
         saved and the hyperparameters that produced the best test error are returned
         """
         best_error = numpy.inf
-        keys = list(self.hyperparameters.keys())
         best_params = dict()
         train, test = self.recommender.split()
         for config in self.get_all_combinations():
             print("running config ")
             print(config)
             self.recommender.set_config(config)
             self.recommender.train()
-            error = self.evaluator.get_rmse(self.recommender.get_predictions(), test)
-            print('Error: %f' % error)
-            if error < best_error:
+            rounded_predictions = ALS.rounded_predictions()
+            test_error = self.evaluator.calculate_recall(test, rounded_predictions)
+            train_error = self.evaluator.calculate_recall(self.recommender.get_ratings(), rounded_predictions)
+            print('Train error: %f, Test error: %f' % (train_error, test_error))
+            if 1 - test_error < best_error:
                 best_params = config
+                best_error = 1 - test_error
+            current_key = self.get_key(config)
+            self.all_errors[current_key] = dict()
+            self.all_errors[current_key]['train_error'] = train_error
+            self.all_errors[current_key]['test_error'] = test_error
         return best_params
 
+    def get_key(self, config):
```

Comment:
Can you add unit tests for this function? and preferably a small example in the documentation.